src/ui/ui_i18n/README.md

```diff
@@ -0,0 +1,325 @@
+# I18n
+
+Kibana relies on several UI frameworks (React and Angular) and
+requires localization in different environments (browser and NodeJS).
+Internationalization engine is framework agnostic and consumable in
+all parts of Kibana (React, Angular and NodeJS). In order to simplify
+internationalization in UI frameworks, the additional abstractions are
+built around the I18n engine: `react-intl` for React and custom
+components for Angular. [React-intl](https://github.com/yahoo/react-intl)
+is built around [intl-messageformat](https://github.com/yahoo/intl-messageformat),
+so both React and Angular frameworks use the same engine and the same
+message syntax.
+
+## Localization files
+
+Localization files have [JSON5](https://github.com/json5/json5) format.
+
+The main benefits of using `JSON5`:
+
+- Objects may have a single trailing comma.
+- Single and multi-line comments are allowed.
+- Additional white space characters are allowed.
+
+Using comments can help to understand which section of the application
+the localization key is used for. Also `namespaces`
+are used in order to simplify message location search. For example, if
+we are going to translate the title of `/management/sections/objects/_objects.html`
+file, we should use message path like this: `'MANAGEMENT.OBJECTS.TITLE'`.
+
+Each Kibana plugin has a separate folder with translation files located at
+```
+src/core_plugins/{plugin_name}/translations/{locale}.json
+```
+
+where `locale` is [ISO 639 language code](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes).
+
+For example:
+```
+src/core_plugins/kibana/translations/en.json
+```
+
+When a new translation file is added, you have to register this file into
+`uiExports.translations` array of plugin constructor parameters. For example:
+```js
+export default function (kibana) {
+  return new kibana.Plugin({
+    uiExports: {
+      translations: [
+        resolve(__dirname, './translations/en.json'),
+      ],
+      ...
+    },
+    ...
+  });
+}
+```
+
+The engine uses a locale resolution process similar to that of the built-in
+Intl APIs to determine which locale data to use based on the `accept-language`
+http header.
+
+The following are the abstract steps i18n engine goes through to resolve the locale value:
+
+- If there's data for the specified locale (localization file is registered in
+  `uiExports.translations`), then that locale will be resolved.
+- If locale data is missing for a leaf locale like `fr-FR`, but there is data
+for one of its ancestors, `fr` in this case, then its ancestor will be used.
+- If `accept-language` header is not presented or previous steps didn't resolve
+the locale, the locale will be resolved to locale defined in `i18n.defaultLocale`
+option at `config/kibana.yml` file.
+- If `i18n.defaultLocale` is not specified or if there is no data for this locale,
+`en` locale will be used.
+
+## I18n engine
+
+I18n engine is the platform agnostic abstractions which helps to supply a locale
```

Comment:
**grammar:** I18n engine is the platform agnostic abstraction that helps to supply locale data ...